lib/ddtrace/propagation/distributed_headers.rb

```diff
@@ -29,11 +30,19 @@ def parent_id
 
     def sampling_priority
       hdr = header(HTTP_HEADER_SAMPLING_PRIORITY)
+
       # It's important to make a difference between no header,
       # and a header defined to zero.
-      return unless hdr
+      return if hdr.nil?
+
+      # Convert header to an integer
       value = hdr.to_i
-      return if value < 0
+
+      # Ensure the parsed number is the same as the original string value
+      # e.g. We want to make sure to throw away `'nan'.to_i == 0`
+      # DEV: Ruby `.to_i` will return `0` if a number could not be parsed
+      return unless value.to_s == hdr.to_s
```

Comment:
This is a bit awkward, but effective. Probably better than using a regex (less expensive I think.)